Use policy for IPCache LPM prefix length generation, not IPCache

joestringer opened the pull request:
Previously, CIDR insertion in policy would queue updates for the IPCache                                                                                                                                             
via the KVstore, then go straight into policy add / policy regeneration         
for endpoints. This would reach into IPCache to fetch the set of prefix         
lengths (ipcache.ToBPFData()) before the original IPCache KVstore update        
has propagated into the local IPCache.                                          
                                                                                
Thus, the set of prefix lengths used in generating the BPF programs may         
not include the prefix lengths for the newly inserted CIDRs. When the           
IPCache KVstore event is eventually processed, no regeneration of BPF is        
triggered so the old set of prefix lengths is used for endpoints until          
the next time the endpoints are regenerated. This leads to improper             
CIDR->Identity mapping in the datapath, which may lead to unexpected            
packet drops as the broader "reserved:world" policy would apply rather          
than the CIDR policy.                                                           
                                                                                
For reference, one plausible solution would be to trigger regeneration          
from IPCache, however this could be costly if you have dozens of IPCache        
updates occurring. There's no good way to batch the updates for these           
currently. Furthermore, it would trigger multiple rebuilds of the               
datapath based on a single policy update.                                       
                                                                                
Alternatively, we should have enough information already during policy          
resolution to generate the appropriate set of prefix lengths for the            
datapath from that policy path. This patch implements this. In other            
words:                                                                          
* Policy generation inserts elements into the L3Policy                          
* Endpoint regeneration uses the L3Policy to determine which CIDR               
prefixes need to be used by the endpoint to perform CIDR->ID mappings           
in order to implement that policy                                               
* The IPCache (by way of the KVstore) will plumb the actual CIDR->ID            
mappings into the BPF maps.                                                     
                                                                                
This should mean that the datapath is eventually consistent with the            
configured CIDR policies.

Note that this will not solve all of the problems; there is still no ability to wait
for the ipcache to be in sync with the policy. However, it does address the
problem that the endpoint BPF programs get stuck in a state where they
can't properly look up the CIDR identity for an extended period.
                                       
Fixes: https://github.com/cilium/cilium/issues/4188
Related: https://github.com/cilium/cilium/issues/4055 (This should further improve test reliability)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/4231)
<!-- Reviewable:end -->

joestringer committed:
endpoint: Use policy for IP LPM, not IPCache

Previously, CIDR insertion in policy would queue updates for the IPCache
via the KVstore, then go straight into policy add / policy regeneration
for endpoints. This would reach into IPCache to fetch the set of prefix
lengths (ipcache.ToBPFData()) before the original IPCache KVstore update
has propagated into the local IPCache.

Thus, the set of prefix lengths used in generating the BPF programs may
not include the prefix lengths for the newly inserted CIDRs. When the
IPCache KVstore event is eventually processed, no regeneration of BPF is
triggered so the old set of prefix lengths is used for endpoints until
the next time the endpoints are regenerated. This leads to improper
CIDR->Identity mapping in the datapath, which may lead to unexpected
packet drops as the broader "reserved:world" policy would apply rather
than the CIDR policy.

For reference, one plausible solution would be to trigger regeneration
from IPCache, however this could be costly if you have dozens of IPCache
updates occurring. There's no good way to batch the updates for these
currently. Furthermore, it would trigger multiple rebuilds of the
datapath based on a single policy update.

Alternatively, we should have enough information already during policy
resolution to generate the appropriate set of prefix lengths for the
datapath from that policy path. This patch implements this. In other
words:
* Policy generation inserts elements into the L3Policy
* Endpoint regeneration uses the L3Policy to determine which CIDR
  prefixes need to be used by the endpoint to perform CIDR->ID mappings
  in order to implement that policy
* The IPCache (by way of the KVstore) will plumb the actual CIDR->ID
  mappings into the BPF maps.

This should mean that the datapath is eventually consistent with the
configured CIDR policies.

Fixes: #4188
Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
policy: Add test for default CIDR prefix lengths

Test that the default set of CIDR prefix lengths looks sane.

Signed-off-by: Joe Stringer <joe@covalent.io>

jrajahalme commented on pkg/endpoint/bpf.go:
```diff
@@ -263,9 +262,12 @@ func (e *Endpoint) writeHeaderfile(prefix string, owner Owner) error {
 	if e.L3Policy != nil {
-		ipv6Ingress, ipv4Ingress := e.L3Policy.Ingress.ToBPFData()
+		// Only egress CIDR policy uses IPCache for now.
+		ipcachePrefixes6, ipcachePrefixes4 = e.L3Policy.Egress.ToBPFData()
```
As this reassigns the prefix length arrays ToBPFData() must also include the default prefix lengths in the returned arrays. Maybe add a comment to that effect here?

jrajahalme reviewed (approved):
LGTM, a minor comment below.

joestringer committed:
policy: Express egress CIDRs in endpoint model

An upcoming commit will make use of the endpoint's L3Policy egress map
to determine which prefix lengths are necessary for generating IPcache
LPM mappings on older Linux kernels. This ensures that the relevant
entries are propagated into this structure for this use.

Over time, the Ingress CIDR L3Policy should do the same (see GH #4129).

Signed-off-by: Joe Stringer <joe@covalent.io>

aanm commented on pkg/policy/l3_test.go:
```diff
@@ -15,26 +15,56 @@
-	m := cidrPolicy.Ingress
+// init prepares the node's allocation range so that the ``ToBPFData()`` calls
+// below can successfully fetch the cluster range.
+func init() {
```
I would prefer that you use the TearUp / TearDown function to set and unset the node allocation range  to make sure the `node.Set..` doesn't affect other tests.

joestringer replied:
I had to introduce a new `Uninitialize()` function in the node package for the teardown. Done this locally, will push out soon.
